lib/ocpp/v201/evse_manager.cpp

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2020 - 2024 Pionix GmbH and Contributors to EVerest
+
+#include <ocpp/v201/evse_manager.hpp>
+
+namespace ocpp {
+namespace v201 {
+
+EvseManager::EvseManager(const std::map<int32_t, int32_t>& evse_connector_structure, DeviceModel& device_model,
+                         std::shared_ptr<DatabaseHandler> database_handler,
+                         std::shared_ptr<ComponentStateManagerInterface> component_state_manager,
+                         const std::function<void(const MeterValue& meter_value, EnhancedTransaction& transaction)>&
+                             transaction_meter_value_req,
+                         const std::function<void(int32_t evse_id)>& pause_charging_callback) {
+    evses.reserve(evse_connector_structure.size());
+    for (const auto& [evse_id, connectors] : evse_connector_structure) {
+        evses.push_back(std::make_unique<Evse>(evse_id, connectors, device_model, database_handler,
+                                               component_state_manager, transaction_meter_value_req,
+                                               pause_charging_callback));
+    }
+}
+
```

Comment:
Dear @marcemmers ,

in general I think it is very fruitful to create some abstract interfaces for the main classes, because testability might be increased and interfaces get more explicit.
Looking at the `EvseManagerInterface`, it is (except for the constructor) a container wrapper.  But despite the `begin/end` functions one could have simply taken a `std::vector`.  To keep things simple, I would also propose to only use a `std::vector` - in case you would like to get some custom iterator behavior, you can still overload `std::begin` and `std::end` to return your custom iterator (as far as I remember this is also the preferred way of customizing containers, as it doesn't couple the logic of how to have a view (iterators) on the container to the container itself).
Note, that this won't work for range-based for loops, because `std::vector` has implemented it's own member functions `begin` and `end` and hence these will be selected while constructing the for loop (see https://en.cppreference.com/w/cpp/language/range-for) - btw. that is exactly the reason why one should prefer the free functions `begin` and `end` vs. member functions.  You could still write a wrapper like `EvseByReference` only for getting your custom iterator and then you could iterate via `for(auto& evse : EvseByReference(evse_map))` ...
An issue, I see with the `EvseManagerInterface` is that it couples the construction of the evse vector to the container itself because this is done in the constructor.  So whenever you pass this container to some other function, you will leak all the construction details.  It might be better here to have a free function, which does exactly the same as the constructor but instead returns the vector.  Furthermore, the `EvseManagerInterface` restricts your flexibility, as you're not able to add/remove evses dynamically (although the `EvseManager` will allow for that by using `get_vector`).
The `is_valid_evse` function for example could also be freed by just accepting the vector and the id and then comparing sizes.  This way, you're decoupling again the responsibility of containing/owning to one class (vector in this case) and the responsibility of creating elements or arguing whether the elements are in a valid state or whatever to another class.  This increases both composability and testability.

Furthermore I see excessive use of `std::shared_ptr` - this is a general issue within libocpp - but as far as I can see, this is not necessary, because the lifetime/ownership relationships are quite clear.  In this case for example, the charge point owns the evses and hence can pass by reference (instead of shared pointer) to the smart charging handler.  Shared pointers would be only necessary if there are valid cases, where the smart charging handler could outlive the charge point - which I think can be ruled out as a valid use case.  The same reasoning applies for most of the other shared pointers.  Although, performance might not be a big issue here, keep in mind that shared pointers need to use locking when constructed/destructed.

Reply:
Hi @a-w50, thanks for the extensive feedback!

I can see what you mean about me increasing the coupling by forcing the use of vectors for whoever implements the `EvseManagerInterface`. Initially I had it split into a pure interface and then a derived implementation, after that changed it into this state but I agree I shouldn't have done that.

I still think this interface class is better then using a vector or a map or whatever container since my whole point is that I don't think we should allow adding more evse's dynamically. This is a config that should be applied once at startup and it does not make sense to change it on the fly. Allowing this under the guise of flexibility opens up potential other issues because not all the initialization code might be performed for example. Whatever is used by the derived class should not matter, that is implementation detail and should not leak.

As for the shared ptr, I agree and I will change it to share a reference to the interface instead of a shared_ptr.